design/one-pager-pluggable-secret-stores.md

```diff
@@ -155,46 +148,39 @@ message DeleteOptions {
-Stores still need to satisfy the same [Store] interface today and could live in
-separate repositories. However, they would need to be packaged and served by a
-Secret Store Plugin binary to be used with Crossplane.
+Deployment of a Secret Store Plugin will be similar to the current provider
+deployment, with one exception. We need to generate a server certificate
+during deployment time and mount it to the plugin pod. This is required to enable
+mTLS communication.
 
-### Deployment
+In addition to the server certificates, we will generate a client certificate
+for each provider and store it in a k8s secret. These certificates will be
+generated by the package manager during the package installation.
+
+Since the core Crossplane communicates with the secret stores as well, it will also
+act as a gRPC client to communicate with the plugin. Hence, it will need a client 
+certificate. To provide this certificate, we will create an empty secret with
+the Crossplane helm chart and mount it to the Crossplane pod. The init container
+will then fill the secret with the client certificate similar to how webhook 
+certificates are provided today.
+
+All these certificates will be signed by the cluster's root CA certificate with proper
+subject and DNS names and stored as k8s secrets. Rotation of these certificates can
+be triggered by deleting those secrets and restarting the Crossplane pod. 
 
-Crossplane helm chart will accept an optional container image for a secret
-store plugin. When provided: 
-- A sidecar container will be added to the Crossplane pod and the package
-manager will inject the same sidecar to provider controller pods with proper
-configuration. 
-- The plugin process will create/apply CRDs during startup for the Secret Stores
-it supports.
-
-#### Communication over Unix Domain Sockets
-
-In Unix based systems, two local process can communicate with each other
-over a [Unix Domain Socket] which typically appears as a file on filesystem,
-e.g. `/var/run/docker.sock`. Two containers in the same pod can communicate with
-each other using Unix Domain sockets with a shared volume. However, in some
-sandboxed environments like [gVisor], this [may not work] depending on the
-isolation level.
-
-[Abstract Unix Domain Sockets] on the other hand does not require sharing a path
-on the filesystem and isolated at [network namespace] level. Given that
-containers in the same Pod uses a shared [network namespace], we don't need
-any specific setup and simply be able to communicate with a plugin sidecar over
-a socket like `@crossplane/secret/default.sock`.
+For deploying External Secret Store plugins, we will introduce a new package type
+`Secret Store`. This package type will be similar to the existing `Provider`
+type. It will contain a CRD to configure the secret store plugin (e.g., `VaultConfig`), 
+a k8s service and a deployment. Required RBAC resources will be generated by the
+package manager.
```

Comment:
There are some concerns about bringing `SecretStore` at the same level as `Providers` or `Configurations` as an extension type since it is not directly related to the _Platform API authoring experience_ but rather more about hosting/running the Control plane. For example, in a Platform Configuration package, you can define dependencies to other Configurations or Providers; however, it does not make sense to define a dependency to a Secret Store since compositions do not require specific secret store types, i.e. agnostic to where secrets are stored.

Stating this, I think it makes sense to start with some documented procedure to install secret store plugins and consider defining it as a package type (and automating this procedure with the package manager) after we feel confident about the above concern and see that it is worth the effort.

We can have a helm chart (instead of an XP package) installing the plugin server, service, and other manifests and document the installation steps accordingly.

Reply:
removed relevant part